doc/design/dynamic_rnn.md

```diff
@@ -0,0 +1,209 @@
+  auto hidden = sigmoid(sum);
+  memories[i+1] = sum;
+  outputs_for_each_timestep[i] = sum;
+}
+
+LoDTensor outputs = TimestepsToLoDTensor(outputs_for_each_timestep);
+```
+
+The `Dynamic RNN` in PaddlePaddle Fluid is basically a syntax sugar to compose operators, such as `while`, `split_lod_tensor_to_timesteps`, `restore_lod_tensor_from_timesteps`.
+
+The following of this document will be organized in several sections:
+
+1. Control flow operators.
+1. No Padding Dynamic RNN algorithm.
+1. Data manipulation operators of RNN.
+
+
+## Control flow operators
+
+### WhileOp
+
+The primary control flow operator to implement dynamic RNN is `WhileOp`. The `WhileOp` holds a sub-block. The operators in the sub-block will be executed again and again while the condition is true. 
+
+#### Sub-Block
+The fragment program of a while op and its sub-block is:
+
+```text
+program {
+  block {
+    idx: 0  # main block
+    parent_idx: -1  # -1 means no parent
+    ops: {
+      ...  # ops before while op
+      
+      op {
+        inputs: ...,
+        outputs: ...,
+        type: "while",
+        attrs: {
+          attr {
+            name: 'sub_block',
+            type: 'BlockID',
+            value: 1  # the sub block id of this while op is 1
+          }
+        }
+      }
+      ...  # ops after while op
+    }
+  }
+  
+  block {
+    idx: 1  # the sub_block of while_op
+    parent_idx: 0  # parent of while block is the main block
+    ops: {
+      ... # ops inside while
+    }
+  }
+}
+```
+
+#### inputs
+
+The while operator has two kinds of inputs. They are
+
+* Condition: A bool scalar. When it's False, the While Op will be terminated. Note that this scalar should always be in CPU memory.
+  * The condition variable is in the parent block. However, it should be updated inside the sub-block of while op. Otherwise, it would result to an endless loop. The condition variable will be an output variable of the while operator, too.
+* X: The external inputs variables, which are required by operators inside the block of While Op.
+  * For example, if there is a hidden fully-connected layer in while operator. The input of the fully-connected layer is output of another operator inside the while operator. The input of this fully-connected layer is not the `external` inputs of the while operator. However, weight tensors of this fully-connected layer are external outputs of the while operator.
+
+  
+#### outputs
+
+* Output: The output variables. They are `assigned` or `push_back` by the operators inside the block of While Op.
+    * It is reasonable for `while operator` to `push_back` its output to an array because 1) the while operator is a loop. 2) the output in every timestep should not be overwritten since they will be used in backward.
+    * The condition and other control flow related operator, like `++i` or `i=0`, could be overwritten since they do not need when backwards. The backward control flow operator of `++i` is `--i`.
```

Comment:
they do not need when backwards --> they are not required in backward stage.
The backward control flow operator of `++i` is `--i`. --> The corresponding control flow operator of `++i` in backward stage is `--i`.